File: 0646-Maximum-Length-of-Pair-Chain/src/Solution.java
import java.util.Arrays;

/**
 * @author qybit
 * @create 2020-12-18 15:04
 * @Since 1.0.0
 */
public class Solution {

    public int findLongestChain(int[][] pairs) {
        Arrays.sort(pairs, (o1, o2) -> o1[1] - o2[1]);
        int res = 1, t = pairs[0][1];
        for (int i = 1; i < pairs.length; i++) {
            if (pairs[i][0] > t) {
                res++;
                t = pairs[i][1];
            }
        }
        return res;
    }
}

File: 0646-Maximum-Length-of-Pair-Chain/cpp/solution.cpp
#include "bits/stdc++.h"

using namespace std;


class Solution {
 public:
  int findLongestChain(vector<vector<int>>& pairs) {
    sort(pairs.begin(), pairs.end(),
         [](vector<int>& a, vector<int>& b) { return a[1] < b[1]; });

    int res = 1, ed = pairs[0][1];
    for (auto p : pairs) {
      if (p[0] > ed) {
        res++;
        ed = p[1];
      }
    }
    return res;
  }
};

int main() {
  Solution s;

  vector<vector<int>> arr;
  arr.push_back({1, 2});
  arr.push_back({2, 3});
  arr.push_back({3, 4});

  cout << s.findLongestChain(arr);
  return 0;
}
